core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/AsyncSigningStage.java

```diff
@@ -128,8 +129,17 @@ private <T extends Identity> CompletableFuture<SdkHttpFullRequest> doSraSign(Sdk
     }
 
     private static void updateAsyncRequestBodyInContexts(RequestExecutionContext context, AsyncSignedRequest signedRequest) {
-        AsyncRequestBody newAsyncRequestBody = signedRequest.payload().isPresent() ?
-                                               AsyncRequestBody.fromPublisher(signedRequest.payload().get()) : null;
+        AsyncRequestBody newAsyncRequestBody;
+        if (signedRequest.payload().isPresent()) {
+            Publisher<ByteBuffer> signedPayload = signedRequest.payload().get();
```

Comment:
We currently have some [sonar cloud](https://sonarcloud.io/project/issues?resolved=false&types=BUG&pullRequest=4251&id=aws_aws-sdk-java-v2&open=AYm4BIRB3hbnKbw3XF2W) bugs for using a pattern similar to this one, in general Java does not ensure that 

```java
signedResquest.payload() == singedRequestPayload()
```

i.e., the second call can return a different instance in a multi-threaded environment. I  think that's not possible for this case but sonar cloud does not seem to be able to figure out that the object is immutable. I think is better to acquire a reference for the optional instance locally and avoid the issue which also avoids calling the method twice, as in

```java
Optional<Publisher<ByteBuffer>> optionalPayload = signedRequest.payload()
if (optionalPayload.isPresent()) {
    Publisher<ByteBuffer> signedPayload = optionalPayload.get();
    ⋯
}
```

Reply:
Yeah, I marked the sonar finding as a false positive, because like you say it doesn't know that signedRequest.payload)() is immutable. But your suggestion makes sense, so I'll make that change. Thanks!